daisy/serial_output_watcher.go

```diff
@@ -0,0 +1,165 @@
+// that c has sufficient buffer space to keep up with the expected
+// signal rate. For a channel that is immediately consumed, a buffer of size 1 is
+// sufficient.
+//
+// Watch needs to be called prior to the instance's CreateInstance step being run.
+// Invocations after that point will be ignored.
+//
+// Polling occurs using the shortest interval of all subscribers,
+// and continues until the instance is deleted. It is guaranteed
+// to continue through a reboot. For a reference of instance lifecycles, see:
+//
+// https://cloud.google.com/compute/docs/instances/instance-life-cycle
+type SerialOutputWatcher interface {
+	Watch(instanceName string, port int64, c chan<- string, interval time.Duration)
+	start(instanceName string)
+}
+
+// NewSerialOutputWatcher constructs a new instance of SerialOutputWatcher using the GCP
+// compute API to make periodic polling requests to check for new serial output.
+func NewSerialOutputWatcher(client serialOutputClient, project, zone string) SerialOutputWatcher {
+	return &serialOutputWatcher{
+		subscriptions: map[instanceAndPort][]watch{},
+		client:        client,
+		project:       project,
+		zone:          zone,
+	}
+}
+
+type serialOutputWatcher struct {
+	subscriptions map[instanceAndPort][]watch
+	client        serialOutputClient
+	project, zone string
+}
+
+// serialOutputWatcher assumes that 404s are returned when an instance doesn't
+// exist. The value `deletionNum404s` determines how many subsequent HTTP 404 responses to
+// see prior to deciding the instance has been deleted (assuming the
+// instance was seen earlier).
+//
+// Although a value of 1 is theoretically sufficient, a higher value allows withstanding
+// unexpected intermittent failures.
+//
+// At the time of writing, the following HTTP error codes were seen:
+//  - Prior to creating instance: 404
+//  - Once deletion finished: 404
+//  - Instance stopped, but not deleted: 400
+//  - Serial output successfully returned: 200
+const deletionNum404s = 3
+
+type instanceAndPort struct {
+	instance   string
+	portNumber int64
+}
+
+type watch struct {
+	channel         chan<- string
+	pollingInterval time.Duration
+}
+
+func (s *serialOutputWatcher) Watch(
+	instanceName string, port int64, c chan<- string, interval time.Duration) {
+	sp := instanceAndPort{instanceName, port}
+	s.subscriptions[sp] = append(s.subscriptions[sp], watch{c, interval})
+}
+
+func (s *serialOutputWatcher) start(instanceName string) {
+	for serialPort := range s.subscriptions {
+		if serialPort.instance != instanceName {
+			continue
+		}
+		callbacks := s.subscriptions[serialPort]
+		if len(callbacks) == 0 {
+			panic(fmt.Sprintf("Start called without subscribers for %v", serialPort))
+		}
+		var subscribers []chan<- string
+		pollingFrequency := callbacks[0].pollingInterval
+		for _, cb := range callbacks {
+			subscribers = append(subscribers, cb.channel)
+			if cb.pollingInterval < pollingFrequency {
+				cb.pollingInterval = pollingFrequency
```

Comment:
Related to the above comment : what if a watcher is registered after start? Should the polling interval be updated in `Watch()` as well?